hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyRenameRequestV1.java

```diff
@@ -0,0 +1,289 @@
+
+  private static final Logger LOG =
+          LoggerFactory.getLogger(OMKeyRenameRequestV1.class);
+
+  public OMKeyRenameRequestV1(OMRequest omRequest) {
+    super(omRequest);
+  }
+
+  @Override
+  @SuppressWarnings("methodlength")
+  public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+      long trxnLogIndex, OzoneManagerDoubleBufferHelper omDoubleBufferHelper) {
+
+    RenameKeyRequest renameKeyRequest = getOmRequest().getRenameKeyRequest();
+    KeyArgs keyArgs = renameKeyRequest.getKeyArgs();
+    Map<String, String> auditMap = buildAuditMap(keyArgs, renameKeyRequest);
+
+    String volumeName = keyArgs.getVolumeName();
+    String bucketName = keyArgs.getBucketName();
+    String fromKeyName = keyArgs.getKeyName();
+    String toKeyName = renameKeyRequest.getToKeyName();
+
+    OMMetrics omMetrics = ozoneManager.getMetrics();
+    omMetrics.incNumKeyRenames();
+
+    AuditLogger auditLogger = ozoneManager.getAuditLogger();
+
+    OMResponse.Builder omResponse = OmResponseUtil.getOMResponseBuilder(
+            getOmRequest());
+
+    OMMetadataManager omMetadataManager = ozoneManager.getMetadataManager();
+    boolean acquiredLock = false;
+    OMClientResponse omClientResponse = null;
+    IOException exception = null;
+    OmKeyInfo fromKeyValue;
+    String fromKey = null;
+    Result result;
+    try {
+      if (toKeyName.length() == 0 || fromKeyName.length() == 0) {
+        throw new OMException("Key name is empty",
+                OMException.ResultCodes.INVALID_KEY_NAME);
+      }
+
+      keyArgs = resolveBucketLink(ozoneManager, keyArgs, auditMap);
+      volumeName = keyArgs.getVolumeName();
+      bucketName = keyArgs.getBucketName();
+
+      // check Acls to see if user has access to perform delete operation on
+      // old key and create operation on new key
+      checkKeyAcls(ozoneManager, volumeName, bucketName, fromKeyName,
+              IAccessAuthorizer.ACLType.DELETE, OzoneObj.ResourceType.KEY);
+      checkKeyAcls(ozoneManager, volumeName, bucketName, toKeyName,
+              IAccessAuthorizer.ACLType.CREATE, OzoneObj.ResourceType.KEY);
+
+      acquiredLock = omMetadataManager.getLock().acquireWriteLock(BUCKET_LOCK,
+              volumeName, bucketName);
+
+      // Validate bucket and volume exists or not.
+      validateBucketAndVolume(omMetadataManager, volumeName, bucketName);
+
+      // Check if toKey exists
+      OzoneFileStatus fromKeyFileStatus =
+              OMFileRequest.getOMKeyInfoIfExists(omMetadataManager, volumeName,
+                      bucketName, fromKeyName, 0);
+      // fromKeyName should exist
+      if (fromKeyFileStatus == null) {
+        // TODO: Add support for renaming open key
+        throw new OMException("Key not found " + fromKey, KEY_NOT_FOUND);
+      }
+
+      // source exists
+      fromKeyValue = fromKeyFileStatus.getKeyInfo();
+      boolean isRenameDirectory = fromKeyFileStatus.isDirectory();
```

Comment:
Thanks @bharatviswa504 for the review help. Hope we have unit test to verify this behavior. Could you please point me to that.

Adding a fair guess, will FS semantics flag help us to cover this case ?

Reply:
FS Semantics flag do you mean ozone.filesystem.paths.enable.

my question was when key is created via ObjectStore API like KeyCreate and rename is called what happens. (When above flag is false)

Because previously from Object Store API rename on directories is not possible. But with this approach how this will work, and how the rename will work?

Reply:
>Because previously from Object Store API rename on directories is not possible

"previously" - you meant before HDDS-2939 ? or are we allowing rename when `ozone.filesystem.paths.enable=true or false` ?

Sorry for multiple questions. Appreciate if you could add an example case that would help me to think better:-). Thanks!

Reply:
ozone.filesystem.paths.enable=false
when key is created like /a/b/c/d i think we don't create intermediate directories right?
So, in that case, how rename works is my question.

Because we check key exists or not by checking dirTable for parents and for leaf FileTable.

Reply:
Thanks @bharatviswa504 for the clear explanation. 
Yes, it won't create intermediate dirs if `ozone.filesystem.paths.enable=false`.

Adding one more point to consider.
[Reference earlier discussion](https://github.com/apache/ozone/pull/1473#discussion_r502350594)

V1 feature version : Following ops shouldn't depend on enableFSPaths flag
```
   File/DirDelete -> Look into fileTable and dirTable for the keys.
   File/DirRename-> Look into fileTable and dirTable for the keys.
```

V1 feature version & enableFSPaths=true
```
   KeyDelete ---> Look into fileTable and dirTable for the keys.
   KeyRename -> supported only in ozone shell. It should look into fileTable and dirTable for the keys.
```
Now the challenge is to support rename for keys as the rename implementation at OM side don't have any clue about whether operation coming from shell objectstore rename or FSRename, right?

To unblock current rename PR and proceed to finish OFS implementation - As we planned earlier, at this initial stage how about target only FS apis and incrementally we can provide supporting ObjectStore KeyCreate, list, delete and rename ops ? I will raise a specific jira issues to handle the Key cases.

Does that make sense to you?